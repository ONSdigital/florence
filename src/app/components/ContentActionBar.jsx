import React from "react";

import PropTypes from "prop-types";

const buttonStyle = {
    STANDARD: "standard",
    POSITIVE: "positive",
    PRIMARY: "primary",
    WARNING: "warning",
    SUBTLE: "subtle",
    INVERT_PRIMARY: "invert-primary"
};
const buttonStyleArray = Object.values(buttonStyle);
const propTypes = {
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            text: PropTypes.string.isRequired,
            interactionCallback: PropTypes.func,
            link: PropTypes.string,
            style: PropTypes.oneOf(buttonStyleArray),
            disabled: PropTypes.bool
        })
    ),
    cancelCallback: PropTypes.func,
    cancelDisabled: PropTypes.bool,
    stickToBottom: PropTypes.bool,
    unsavedChanges: PropTypes.bool
};

const ContentActionBar = props => {
    const outerContainerClasses = props.stickToBottom
        ? "grid grid--justify-center content-action-bar content-action-bar--bottom-fixed"
        : "content-action-bar";
    const innerContainerClasses = props.stickToBottom ? "grid__col-10" : "";
    const contentButtons = props.buttons.map((btn, index) => {
        let classes = "btn";
        if (btn.style != null && btn.style !== buttonStyle.STANDARD) {
            classes += ` btn--${btn.style}`;
        }
        classes += index > 0 ? " btn--margin-right btn--margin-left" : "";
        if (btn.interactionCallback != null) {
            return (
                <button id={btn.id} type="button" key={btn.id}onClick={btn.interactionCallback} disabled={btn.disabled} className={classes}>
                    {btn.text}
                </button>
            );
        } else {
            return (
                <a id={btn.id} href={btn.link} className={classes} key={btn.id}>
                    {btn.text}
                </a>
            );
        }
    });
    const unsavedChangesWarning = (
        <span className="content-action-bar__warn">
            <svg className="svg-icon--action-bar" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" focusable="false" role="img" aria-label="warning icon">
                <path
                    d="M256,34.297L0,477.703h512L256,34.297z M256,422.05c-9.22,0-16.696-7.475-16.696-16.696s7.475-16.696,16.696-16.696
                            c9.22,0,16.696,7.475,16.696,16.696S265.22,422.05,256,422.05z M239.304,344.137V177.181h33.391v166.956H239.304z"
                />
            </svg>
            <span className="content-action-bar__warn__text"> You have unsaved changes</span>
        </span>
    );
    return (
        <div className={outerContainerClasses}>
            <div className={innerContainerClasses}>
                <div className="padding-bottom--1 padding-top--1">
                    {contentButtons}
                    {props.cancelCallback && (
                        <button
                            id="btn-cancel"
                            type="button"
                            onClick={props.cancelCallback}
                            disabled={props.cancelDisabled}
                            className="btn btn--invert-primary btn--margin-left"
                        >
                            Cancel
                        </button>
                    )}
                    {props.unsavedChanges && unsavedChangesWarning}
                </div>
            </div>
        </div>
    );
};
ContentActionBar.propTypes = propTypes;
export default ContentActionBar;
