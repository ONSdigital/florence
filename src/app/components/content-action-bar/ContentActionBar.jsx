import React from "react";

import PropTypes from "prop-types";
import Warning from "../../icons/Warning";

const propTypes = {
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            text: PropTypes.string.isRequired,
            interactionCallback: PropTypes.func,
            link: PropTypes.string,
            style: PropTypes.oneOf(["standard", "positive", "primary", "warning", "subtle", "invert-primary"]),
            disabled: PropTypes.bool
        })
    ),
    cancelCallback: PropTypes.func,
    cancelDisabled: PropTypes.bool,
    stickToBottom: PropTypes.bool,
    unsavedChanges: PropTypes.bool
};

const ContentActionBar = props => {
    const contentButtons = props.buttons.map((btn) => {
        let classes = "btn btn--margin-right";
        if (btn.style != null && btn.style !== "standard") {
            classes += ` btn--${btn.style}`;
        }
        if (btn.interactionCallback != null) {
            return (
                <button id={btn.id} type="button" key={btn.id} onClick={btn.interactionCallback} disabled={btn.disabled}
                        className={classes}>
                    {btn.text}
                </button>
            );
        }
        return (
            <a id={btn.id} href={btn.link} className={classes} key={btn.id}>
                {btn.text}
            </a>
        );
    });
    const unsavedChangesWarning = (
        <span className="content-action-bar__warn">
            <Warning classes="svg-icon--action-bar" ariaLabel="warning icon" viewBox="0 0 512 512"/>
            <span className="content-action-bar__warn__text"> You have unsaved changes</span>
        </span>
    );
    return (
        <div className={`content-action-bar ${props.stickToBottom ? "content-action-bar--bottom-fixed" : ""}`}>
                {contentButtons}
                {props.cancelCallback && (
                    <button
                        id="btn-cancel"
                        type="button"
                        onClick={props.cancelCallback}
                        disabled={props.cancelDisabled}
                        className="btn btn--invert-primary"
                    >
                        Cancel
                    </button>
                )}
                {props.unsavedChanges && unsavedChangesWarning}
        </div>
    );
};
ContentActionBar.propTypes = propTypes;
export default ContentActionBar;
