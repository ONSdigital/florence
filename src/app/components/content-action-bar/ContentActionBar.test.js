import React from "react";
import ContentActionBar from "./ContentActionBar.jsx";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

describe("ContentActionBar", () => {
    describe("is created stuck at bottom with a single action button, cancel and with unsaved changes", () => {
        const contentActionBarProps = {
            buttons: [
                {
                    id: "foo-id",
                    text: "foo",
                    interactionCallback: () => {},
                    style: "positive",
                    disabled: false
                }
            ],
            cancelCallback: () => {},
            stickToBottom: true,
            unsavedChanges: true
        };

        const component = mount(<ContentActionBar {...contentActionBarProps} />);
        it("should display 2 buttons, be attached at the bottom and have a warning that there are unsaved changes", () => {
            expect(component.find(".content-action-bar--bottom-fixed").length).toBe(1);
            expect(component.find(".btn").length).toBe(2);
            expect(component.find("#btn-cancel").exists()).toBe(true);
            const fooButton = component.find("#foo-id");
            expect(fooButton.exists()).toBe(true);
            expect(fooButton.hasClass("btn--positive")).toEqual(true);
            expect(component.find(".content-action-bar__warn").exists()).toBe(true);
        });
        const componentForSnapshot = renderer.create(<ContentActionBar {...contentActionBarProps} />);
        expect(componentForSnapshot.toJSON()).toMatchSnapshot();
    });
    describe("is created floating with two links", () => {
        const contentActionBarProps = {
            buttons: [
                {
                    id: "bar-id",
                    text: "bar",
                    link: "/",
                    style: "warning",
                    disabled: false
                },
                {
                    id: "baz-id",
                    text: "baz",
                    link: "/",
                    style: "primary",
                    disabled: false
                }
            ],
            stickToBottom: false,
            unsavedChanges: false
        };
        const component = mount(<ContentActionBar {...contentActionBarProps} />);
        it("should display 2 anchors and be floating", () => {
            expect(component.find(".content-action-bar--bottom-fixed").length).toBe(0);
            expect(component.find(".btn").length).toBe(2);
            expect(component.find("a").length).toBe(2);
            expect(component.find("#btn-cancel").exists()).toBe(false);
            const barButton = component.find("#bar-id");
            const bazButton = component.find("#baz-id");
            expect(barButton.exists()).toBe(true);
            expect(bazButton.exists()).toBe(true);
            expect(barButton.hasClass("btn--warning")).toEqual(true);
            expect(bazButton.hasClass("btn--primary")).toEqual(true);
            expect(component.find("#baz-id").exists()).toBe(true);
            expect(component.find(".content-action-bar__warn").exists()).toBe(false);
        });
        const componentForSnapshot = renderer.create(<ContentActionBar {...contentActionBarProps} />);
        expect(componentForSnapshot.toJSON()).toMatchSnapshot();
    });
});
