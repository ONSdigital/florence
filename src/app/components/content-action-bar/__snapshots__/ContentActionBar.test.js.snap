// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentActionBar with a different set of props passed to it matches the snapshot 1`] = `
<div
  className="content-action-bar "
>
  <a
    className="btn btn--margin-right btn--warning"
    href="/"
    id="bar-id"
  >
    bar
  </a>
  <a
    className="btn btn--margin-right btn--primary"
    href="/"
    id="baz-id"
  >
    baz
  </a>
</div>
`;

exports[`ContentActionBar with a set of props passed to it matches the snapshot 1`] = `
<div
  className="content-action-bar content-action-bar--bottom-fixed"
>
  <button
    className="btn btn--margin-right btn--positive"
    disabled={false}
    id="foo-id"
    onClick={[Function]}
    type="button"
  >
    foo
  </button>
  <button
    className="btn btn--invert-primary"
    id="btn-cancel"
    onClick={[Function]}
    type="button"
  >
    Cancel
  </button>
  <span
    className="content-action-bar__warn"
  >
    <svg
      aria-label="warning icon"
      className="svg-icon--action-bar"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256,34.297L0,477.703h512L256,34.297z M256,422.05c-9.22,0-16.696-7.475-16.696-16.696s7.475-16.696,16.696-16.696 c9.22,0,16.696,7.475,16.696,16.696S265.22,422.05,256,422.05z M239.304,344.137V177.181h33.391v166.956H239.304z"
      />
    </svg>
    <span
      className="content-action-bar__warn__text"
    >
       You have unsaved changes
    </span>
  </span>
</div>
`;
